events/events_network.c

```diff
@@ -292,30 +323,47 @@ struct eventrec *
 events_network_get(void)
 {
 	struct eventrec * r;
-	size_t nfds = socketlist_getsize(S);
 
 	/* We haven't found any events yet. */
 	r = NULL;
 
+	/* We have no fds, thus we have no events. */
+	if (fds == NULL)
+		return (r);
+
 	/* Scan through the fd sets looking for ready sockets. */
 	for (; fdscanpos < nfds; fdscanpos++) {
+		struct pollfd * current = &fds[fdscanpos];
+
+		/* File descriptor was closed, ignore it in future polls. */
+		if (current->revents & POLLNVAL) {
+			current->fd = -1;
+			continue;
+		}
+
 		/* Are we ready for reading? */
-		if (FD_ISSET(fdscanpos, &readfds)) {
-			r = socketlist_get(S, fdscanpos)->reader;
-			socketlist_get(S, fdscanpos)->reader = NULL;
+		if (current->revents & POLLIN) {
+			r = socketlist_get(S, (size_t)current->fd)->reader;
```

Comment:
If `r` is NULL here, we should keep scanning.  This can happen if _cancel has been called, since we're no longer clearing the socket readable/writable bits.